### PR TITLE
Op's Sign-aware Value Predemotion Truncate

### PR DESCRIPTION
**Pull request by raikonenfnu**
Implements a pass that checks an op is signed, and has constant
operands. If condition is fufilled, then will "predemote" the value into
the value of signed target type while keeping the source type. This is
to ensure correct demotion of value based on the signedness of ops.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp [hunk 1]
```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
...
+  if (newValue == value) {
+    return operand;
+  }
+  Value truncatedConst = rewriter.create<arith::ConstantOp>(
+      constOp->getLoc(), IntegerAttr::get(oldType, newValue));
+  return truncatedConst;
+}
+
+// Utility function to check that op is signed or not.
+bool isSignedOp(Operation *op) {
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:242]: This is fine for now I guess. But I'd think we need to define op traits in the upstream so that we don't need to maintain a list of ops here specifically by ourselves. That is, extend the [definitions in SPIR-V dialect](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/SPIRV/IR/SPIRVOpTraits.h#L23-L27) to all dialects. Let me create a patch for that.

**Comment by powderluv** [comment]
https://discord.com/channels/689900678990135345/982385246475522098/982525491607130173

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp [hunk 2]
```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
...
+                                        this->getTypeConverter());
+    Value rhs = truncateOutOfRangeValue(op.getOperand(1), rewriter,
+                                        this->getTypeConverter());
+    if (lhs == op.getOperand(0) && rhs == op.getOperand(1)) return failure();
+    rewriter.replaceOpWithNewOp<arith::CmpIOp>(op, cmpPredicate, lhs, rhs);
+    return success();
+  }
+};
+
+struct ConvertSignSensitiveArithSelectOp
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:289]: Do we need two separate classes here? What about a single pattern matching everything to handle both? Something like:

```c++
struct ConvertSignSensitiveArithOp: public OpConversionPattern {
    ConvertSignSensitiveArithOp(MLIRContext *context, TypeConverter &typeConverter)
      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*patternBenefit=*/1, context) {}

  LogicalResult matchAndRewrite(
      Operation *op, ArrayRef<Value> operands,
      ConversionPatternRewriter &rewriter) const override {
        if (!isSignedOp(op)) return failure();
        SmallVector<Value, 4> newOperands;
        bool truncated = false;
        for (Value oldOperand: op->getOperands()) {
          Value newOperand = truncateOutOfRangeValue(...);
          newOperands.push_back(newOperand);
          truncated |= (oldOperand != newOperand);
        }
        if (!truncated) return failure();
        OperationState state(op->getLoc(), op->getName().getStringRef(), newOperands,
                                             op->getResultTypes(), op->getAttrs());
        Operation *newOp = rewriter.create(state);
        rewriter.replaceOp(op, newOp->getResults());
        return success();
      }
};
```

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp [hunk 3]
```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
   }
 };
 
+// Truncates input operand if it is a constant and is out of range of target
+// type.
+Value truncateOutOfRangeValue(Value operand,
+                              ConversionPatternRewriter &rewriter,
+                              TypeConverter *typeConverter) {
+  if (!operand.getDefiningOp()) {
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:218]: `auto constOp = operand.getDefininingOp<arith::ConstantOp>()` here.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp [hunk 4]
```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
...
+  auto oldType = operand.getType();
+  auto newType = typeConverter->convertType(oldType);
+  if (oldType == newType) {
+    return operand;
+  }
+  auto constOp = dyn_cast<arith::ConstantOp>(operand.getDefiningOp());
+  if (!constOp) {
+    return operand;
+  }
+  auto intAttr = constOp->getAttr("value").dyn_cast<IntegerAttr>();
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:230]: `constOp.getValue().dyn_cast<IntegerAttr>()`

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp [hunk 5]
```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
...
+  APInt value = intAttr.getValue();
+  APInt newValue = value.truncSSat(newType.getIntOrFloatBitWidth());
+  if (newValue == value) {
+    return operand;
+  }
+  Value truncatedConst = rewriter.create<arith::ConstantOp>(
+      constOp->getLoc(), IntegerAttr::get(oldType, newValue));
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:237]: Why this is still using `oldType` here?

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp [hunk 6]
```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
   }
 };
 
+// Truncates input operand if it is a constant and is out of range of target
+// type.
```
Comment (antiagainst) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:214]: .. Otherwise, returns the original operand.

**Review by antiagainst — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp (hunk 1 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Sounds good, added this note!

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp (hunk 2 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Done! thanks :)

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp (hunk 3 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Thanks! was about to use dyn_cast_or_null, but this is much better :)

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp (hunk 4 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: done!

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp (hunk 5 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Done! this was legacy code, since it used to be a different pass, it just demoted the value and wait for ConvertType to actually demote the type. Also it used to have issue with other operand type not being demoted yet causing error.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp (hunk 6 above)
Comment (raikonenfnu) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp]: Done!

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Have a little concern over this one. (other folks who know this can correct me), but this seems to be correcting the IR after the fact, i.e. the error happened somewhere else, and correcting at that stage would be "better defined". Such post-fact fixes can be foot guns. Let me elaborate a bit.

There are two way you can represent signedness
1) The type can carry sign information
2) The type can be unsigned (so all values are bag of bits), but the operation can carry the signedness information.

LLVM does the latter. In MLIR dialects have a choice what they want to implement. MHLO for example carries signedness with the type. Arith dialect (and other dialects in MLIR core) are consistent with LLVM with the op carrying signedness. For the MHLO path it is the conversion from MHLO -> Linalg that handles the signedness appropriately, i.e. converts a program where signedness is carried by the type into a program where signedness is carried by the operation. 

There are two places where the "correct" fix would be
1) If Torch dialect carries signedness on the type, conversion from Torch -> Linalg should do what MHLO -> Linalg does.
2) If Torch itself carries signedness on the op, then lowering into torch should handle this.

TL;DR I think this is a case of wrong input IR, and fix for this within IREE core looks like the wrong place to do it.

**Comment by raikonenfnu** [comment]
@MaheshRavishankar We can take a look at the torch-mlir -> linalg conversion and see if there exist hints/traits/attributes for the signedness there. May lead to cleaner solution. We have this pass in our shark-fork anyways so it's not completely blocking for now and we can spend more time thinking about the best solution. 

If we do end up with the fix in frontend(torch-mlir), I was thinking it would be good to have warning in IREE when we demote signless to unsigned since it may not be accurate. I think this will help debug for future pass writers.

Also, any thoughts of changing default demotion to signed instead of unsigned? (since many of our use cases are probably signed anyways)

**Comment by MaheshRavishankar** [comment]
> @MaheshRavishankar We can take a look at the torch-mlir -> linalg conversion and see if there exist hints/traits/attributes for the signedness there. May lead to cleaner solution. We have this pass in our shark-fork anyways so it's not completely blocking for now and we can spend more time thinking about the best solution.
> 
> If we do end up with the fix in frontend(torch-mlir), I was thinking it would be good to have warning in IREE when we demote signless to unsigned since it may not be accurate. I think this will help debug for future pass writers.
> 

Since IREE core sees the IR only after Linalg, it doesnt have enough information to flag this. It is assumed that all the signedness related stuff has been resolved correctly during conversion to Linalg. 

> Also, any thoughts of changing default demotion to signed instead of unsigned? (since many of our use cases are probably signed anyways)

There might be bugs, but default is neither signed nor unsigned. Its a bag of bits...

**Comment by antiagainst** [comment]
I think it's still useful to have this fix (and the demotion passes) within IREE. Here the issue is that we have a bug in these demotion patterns. What are given from the frontend are consistent actually (although they are using 64-bit values, which is not good)--it contains a signless 64-bit constant value and it uses op-carried-signess to interpret that. However in our demotion patterns we aren't considering that when changing the bitwdith; instead we just forcefully treat all signless constants in one single way. That's a bug to be fixed in the demotion to me.

Though I agree that it would be nice if we can teach the frontends to use 32-bit values for all the goodness, but as discussed in other threads previously, it might not be always possible. Such demotion patterns in IREE are useful in the meantime--sometimes what are generated from frontend are not easy to fix / take time to fix.

**Comment by benvanik** [comment]
The real issue is that even if the semantics are respected it still won't always be right. Stella and I have a tale of a model internally that used FLT_MAX*0.7 as a sentinel and as you can imagine things weren't happy with that being narrowed to f32 :P
The best solution as you mention is to use the right types in the input, and everything else is squarely in YMMV territory. We should fix up what we can but otherwise try not to spend time chasing things here.

**Comment by MaheshRavishankar** [comment]
> I think it's still useful to have this fix (and the demotion passes) within IREE. Here the issue is that we have a bug in these demotion patterns. What are given from the frontend are consistent actually (although they are using 64-bit values, which is not good)--it contains a signless 64-bit constant value and it uses op-carried-signess to interpret that. However in our demotion patterns we aren't considering that when changing the bitwdith; instead we just forcefully treat all signless constants in one single way. That's a bug to be fixed in the demotion to me.
> 

Yes it is a bug in demotion, but by the time it gets to IREE (depends on what you call IREE. I am seeing this as "post-MHLO") how will we know where to apply this. With respect to core IREE compiler it is garbage in garbage out. We cant realistically fix this retro-actively in IREE core (and dont think we actually can). So where-ever we are converting from sign carried on type to sign carried on op is where this should be fixed.

**Comment by benvanik** [comment]
Closing stale PR.
